### PR TITLE
Optional half matmul

### PR DESCRIPTION
**Pull request by Eliulm**

### CODE REVIEW

File: examples/mlperf/initializers.py [hunk 1]
```diff
@@ -41,7 +41,8 @@ def __init__(self, in_features, out_features, bias=True, std=0.02):
     self.bias = Tensor.zeros(out_features, dtype=dtypes.float32) if bias else None
   
   def __call__(self, x:Tensor):
-    return x.linear(self.weight.cast(dtypes.default_float).transpose(), self.bias.cast(dtypes.default_float) if self.bias is not None else None)
+    x = x.contiguous().cast(dtypes.half if getenv("HALF_LINEAR", 0) else dtypes.default_float)
```
Comment (chenyuxyz) [inline examples/mlperf/initializers.py:44]: can you make a variable for `dtypes.half if getenv("HALF_LINEAR", 0) else dtypes.default_float`? i think that reads better and less repeated code

File: examples/mlperf/model_train.py [hunk 2]
```diff
@@ -430,6 +430,7 @@ def train_bert():
   target, achieved                                      = getenv("TARGET", 0.72), False
 
   config["DEFAULT_FLOAT"] = dtypes.default_float.name
+  config["HALF_LINEAR"]   = getenv("HALF_LINEAR", 0)
```
Comment (chenyuxyz) [inline examples/mlperf/model_train.py:433]: would be nice to also log dropout so it's easy to tell if dropout is used or not for now, not related to this pr though

File: examples/mlperf/initializers.py [hunk 3]
```diff
@@ -41,7 +41,8 @@ def __init__(self, in_features, out_features, bias=True, std=0.02):
     self.bias = Tensor.zeros(out_features, dtype=dtypes.float32) if bias else None
   
   def __call__(self, x:Tensor):
-    return x.linear(self.weight.cast(dtypes.default_float).transpose(), self.bias.cast(dtypes.default_float) if self.bias is not None else None)
+    matmul_dtype = dtypes.half if getenv("HALF_LINEAR", 0) else dtypes.default_float
+    return x.contiguous().cast(matmul_dtype).linear(self.weight.contiguous().cast(matmul_dtype).transpose(), self.bias.cast(dtypes.default_float) if self.bias is not None else None)
```
Comment (chenyuxyz) [inline examples/mlperf/initializers.py:45]: do you need these `contiguous`? weight should be contiguous and is there a benefit to do this with `x`?

**Review by chenyuxyz — COMMENTED** [review]
have you benchmarked default float32 with / without this cast? curious to know the speed difference

File: examples/mlperf/initializers.py (hunk 3 above)
Comment (Eliulm) [reply examples/mlperf/initializers.py:45]: currently, uop count explodes when I don't do that. Seems like a bug. I think Qazalin is on that already

Comment (chenyuxyz) [reply examples/mlperf/initializers.py:45]: if it's a temporary issue, can you add a TODO note for some context

Comment (Qazalin) [reply examples/mlperf/initializers.py:46]: Yeah I'm looking into the root cause for this

**Comment by Eliulm** [comment]
> have you benchmarked default float32 with / without this cast? curious to know the speed difference

So on my machine, full float is 10.8 TFLOPS, float with half linear is 19TFLOPS and full half is 20.8TFLOPS

**Comment by chenyuxyz** [comment]
Cool merging! Interesting that changing attention matmul caused the nan

**Pull request merged**